sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -472,6 +525,75 @@ void Scheduler::cleanupCommands(const std::vector<Command *> &Cmds) {
   }
 }
 
+void Scheduler::deferMemObjRelease(const std::shared_ptr<SYCLMemObjI> &MemObj) {
+  {
+    std::lock_guard<std::mutex> Lock{MDeferredMemReleaseMutex};
+    MDeferredMemObjRelease.push_back(MemObj);
+  }
+  cleanupDeferredMemObjects(false);
+}
+
+inline bool Scheduler::isNoDeferredMemObjects() {
+  std::lock_guard<std::mutex> Lock{MDeferredMemReleaseMutex};
+  return MDeferredMemObjRelease.empty();
+}
+
+void Scheduler::cleanupDeferredMemObjects(bool ForceWait) {
+  if (isNoDeferredMemObjects())
+    return;
+
+  // Need to aggregate ready to release object to acquire write lock once.
+  std::list<std::shared_ptr<SYCLMemObjI>> ObjsReadyToRelease;
+  {
+    ReadLockT Lock(MGraphLock, std::try_to_lock);
+    // if we need blocking mode - force lock waiting
+    if (!Lock.owns_lock() && ForceWait)
+      Lock.lock();
+    if (Lock.owns_lock()) {
+      // Not expected that ForceWait == true with be used in parallel with
+      // adding MemObj to storage, no such scenario.
+      std::lock_guard<std::mutex> LockDef{MDeferredMemReleaseMutex};
+      auto MemObjIt = MDeferredMemObjRelease.begin();
+      while (MemObjIt != MDeferredMemObjRelease.end()) {
+        MemObjRecord *Record = MGraphBuilder.getMemObjRecord((*MemObjIt).get());
+        if (!Record) {
+          // Just trigger delete since no operations on object was perfromed and
+          // no commands and other to wait for
+          MemObjIt = MDeferredMemObjRelease.erase(MemObjIt);
+          continue;
+        }
+        if (!waitForRecordToFinish(Record, Lock, ForceWait)) {
+          MemObjIt++;
+          continue;
+        }
+        ObjsReadyToRelease.push_back(*MemObjIt);
+        MemObjIt = MDeferredMemObjRelease.erase(MemObjIt);
+      }
+    }
+  }
+  if (ObjsReadyToRelease.empty())
+    return;
+
+  std::vector<std::shared_ptr<stream_impl>> StreamsToDeallocate;
+  std::vector<std::shared_ptr<const void>> AuxResourcesToDeallocate;
+  {
+    WriteLockT Lock(MGraphLock, std::try_to_lock);
+    // if we need blocking mode - force lock waiting
+    if (!Lock.owns_lock() && ForceWait)
+      acquireWriteLock(Lock);
+    if (Lock.owns_lock()) {
+      for (auto &MemObj : ObjsReadyToRelease)
```

Comment:
```suggestion
      for (std::shared_ptr<SYCLMemObjI>> &MemObj : ObjsReadyToRelease)
```
Please, do not use `auto` where the type is not immediately clear.

Reply:
fixed here https://github.com/intel/llvm/pull/6837/commits/b5e85de484fb026152cfcaa8d3dd8bbc79c158d3